inf-ruby.el

```diff
@@ -683,6 +675,48 @@ This function also removes itself from `pre-command-hook'."
                 (point)))))
         (buffer-substring-no-properties (point) (line-end-position))))))
 
+(defun ruby-shell--encode-string (string)
+  "Escape all backslashes, single quotes, and newlines in STRING."
+  (cl-reduce (lambda (string subst)
+               (replace-regexp-in-string (car subst) (cdr subst) string))
+             '(("\\\\" . "\\\\\\\\")
+               ("'" . "\\\\'")
+               ("\n" . "'\"\\\\n\"'"))
```

Comment:
I wanted to comment on this unexpected choice of escaping, but I see you're taking advantage of both the relatively obscure string concatenation syntax, and the fact that almost no escapes work in single quotes (meaning, this list gets to be short). An extra comment for our descendants wouldn't hurt, though.

Reply:
Yea I wasn't really sure the best way to go about this. python-mode is just using `json-serialize` on emacs >=28 and on earlier emacs does `(require 'json)` and `json-encode-string`, which produces a new double quoted string with newlines and everything escaped for python. But for ruby we would need to then also escape # so unwanted string interpolation doesn't happen. So I figured if I'm escaping one character anyway it's simpler to just escape three characters and use single quoted string rather than do json encoding plus escaping #.
`s-replace-all` from s.el does exactly what we want, taking an alist of pairs to replace, but I didn't want to add a dependency even if nowadays almost everyone probably has s.el. Built into emacs since 28 there is `string-replace` which would be nicer but didn't want to require new emacs just for this, so `cl-reduce` on `replace-regexp-in-string` was best I came up with, needing a ridiculous number of backslashes.

Rather than the obscure string concatenation '"\n"' and comment explaining it, can we just use double quoted strings? I think it's still just needed to escape a few characters \ " newline, and # but I wasn't quite as confident as with single quoted.

Also I just realized in inf-ruby there is already an almost identical `ruby-escape-single-quoted`, used on lines 927-928 for this same purpose of escaping a string that is sent to inf-ruby-proc between single quotes, except I think it's slightly wrong. It replaces newlines with a literal \n. And it doesn't escape double quotes, which isn't needed but the docstring says it does.

Reply:
I read the ruby reference on string literals and I'm pretty confident it's correct using double quoted strings just escaping backslash, #, and ", so I switched it to double quoted strings so it can use \n for newline and not need that weird concating a double quoted string with \n in the middle of single quoted string.

There's still the question of the already existing `ruby-escape-single-quoted` function which is replacing newlines with \n in single quoted strings where \n is literally that and not a newline. I think it's wrong but not causing any problems because it's only used at present for completions which can't have newlines anyway.

Reply:
I liked the "obscure concatenation", but using double-quotes is probably easier to read, thank you.

Good point about `ruby-escape-single-quoted`, but it can be updated separately.

Reply:
Should I make a separate issue for that? The only thing is if we want to use the same escaping function for both uses, then I probably shouldn't name this one `ruby-shell--encode-string` but `ruby-escape-double-quoted` or something.